### ISSUE TITLE
C# 9 Analyzer: Flag suspect pattern usages

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn-analyzers#3976]
C# 9 is going to introduce new pattern matching syntax around `or`, `and`, and `not`. There are inevitably going to end up confusing people, and I can already see the flood of questions on the C# discord about "Why does the && not do what I told it to do here?". Some examples where this can fall down:

```cs
switch ((bool?)false)
{
    case true || false: break; // Does not handle false, because false || true is a constant expression!
	case null: break;
}

if (o is 1 || true) {} // Always enters this if
```

We should consider having a rule to flag these types of suspect cases and offer to rewrite them using `or` patterns, much like we do for `!is` today. The basic rules for such an analyzer (I think) have 2 forms:

Form 1: Pattern expressions

These are things inside a pattern, such as in a case block or nested inside some other pattern. This is the much simpler case:
1. If you encounter a `||`, offer to change it to an `or`. If you encounter an `&&`, offer to change it to an `and`.

Inside a pattern, this can almost never go wrong, as everything has to be constant, or the compiler will error anyway. There are cases when you could legitimately use `||` and `&&` inside a pattern (if, say, you were composing a single boolean check out of multiple `#if`-set flags), but these cases are few, far between, confusing to the reader, and I would argue should have a warning suppression.

Form 2: Trailing `is` expression constants

These are much harder to detect and implement, and false positives are much more likely, but I think at least offering an `info` diagnostic and a refactoring would be useful. A implementation of this rule might be:

1. See an `is` expression.
2. Walk to the next node past the pattern (and trivia) of the expression. If that is not a logical operator (either `&&` or `||`), goto step 5.
3. Walk to the next sub-expression, the expression on the right of this logical operator.
4. Check the constant value of this expression. If there is no constant value, goto step 5. If there is, add the expression to a list of constants to add to the pattern, and whether it was an `and` or an `or`. Go to step 2.
5. If no constants were found, end. Otherwise, gather the list of constants and their conjunctions, and rewrite the pattern use `and`s and `or`s instead of `&&` and `||`.

Finally, we should consider having a refactoring going the other direction, particularly for non-constants. For example, if I write `if (a is nonConstant1 or nonConstant2)`, I'd get an error. A refactoring to convert that to `if (a == nonConstant1 || a == nonConstant2)` would be useful. For this idea, also consider saving `a` to a temp if it's something like a method call, as that's what the pattern form will do. Perhaps offer 2 refactorings, one that will save and one that will not.

Both of these refactorings might change the semantics of the code, so it's important to call that out. But I believe that new users will end up being confused by the different syntaxes here, and we should try to help them.

**Comment by mavasani** [comment]
Thanks @333fred. Analyzers related to new language constructs are generally implemented in Roslyn repo as IDExxxx rules. CAxxxx rules in this repo are more for API analyzers + generic code quality suggestions. Likely need to transfer this item to Roslyn. Tagging @sharwell @CyrusNajmabadi

**Comment by 333fred** [comment]
I talked with @CyrusNajmabadi about this before filing, he suggested that it should be an analyzer, not an IDExxxx rule, so I opened here. Happy to move it over though :).

**Comment by mavasani** [comment]
Ah okay. I trust Cyrus' call here :). Cyrus, does this mean we also port a similar IDE analyzer for `!is` you authored in Roslyn to this repo: https://github.com/dotnet/roslyn/pull/44878

**Comment by mavasani** [comment]
@333fred By the way, we still cannot author refactorings in a completely reasonable fashion in this repo. Code refactorings in Roslyn rely on internal APIs, which are important for an ideal implementation. If the refactoring are just the opposite of analyzers, then it would be weird to have them in a separate repo then the analyzer. I am also wondering which analyzer package would we put such language-specific analyzers, there are no rules or analyzers for language specific suggestions in this repo, so we would need to identify that.

@sharwell @CyrusNajmabadi should this be discussed at a design meeting?
